presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraClusteringPredicatesExtractor.java

```diff
@@ -87,75 +64,148 @@ public CassandraClusteringPredicatesExtractor(List<CassandraColumnHandle> cluste
                         }
-                        return columnValues.build();
-                    },
-                    discreteValues -> {
-                        if (discreteValues.isWhiteList()) {
-                            return ImmutableSet.copyOf(discreteValues.getValues());
+                    }
+
+                    if (singleValues.size() == 1) {
+                        predicate = CassandraCqlUtils.validColumnName(columnHandle.getName()) + " = "
+                                + singleValues.get(0);
+                    }
+                    else if (singleValues.size() > 1) {
+                        if (protocolVersion < 4) {
+                            return "Not Supported";
                         }
-                        return ImmutableSet.of();
-                    },
-                    allOrNone -> ImmutableSet.of());
-
-            if (!values.isEmpty()) {
-                clusteringColumnValues.add(values);
-                domainsBuilder.put(columnHandle, domain);
+                        predicate = CassandraCqlUtils.validColumnName(columnHandle.getName()) + " IN ("
+                                + Joiner.on(",").join(singleValues) + ")";
+                    }
+                    else if (!rangeConjuncts.isEmpty()) {
+                        predicate = Joiner.on(" AND ").join(rangeConjuncts);
+                    }
+                    return predicate;
+                }, discreteValues -> {
+                    if (discreteValues.isWhiteList()) {
+                        ImmutableList.Builder<Object> discreteValuesList = ImmutableList.builder();
+                        for (Object discreteValue : discreteValues.getValues()) {
+                            discreteValuesList.add(CassandraCqlUtils.cqlValue(toCQLCompatibleString(discreteValue),
+                                    columnHandle.getCassandraType()));
+                        }
+                        if (protocolVersion < 4) {
+                            return "Not Supported";
+                        }
+                        String predicate = CassandraCqlUtils.validColumnName(columnHandle.getName()) + " IN ("
+                                + Joiner.on(",").join(discreteValuesList.build()) + ")";
+                        return predicate;
+                    }
+                    return null;
+                }, allOrNone -> null);
+
+                if (predicateString != null) {
+                    // Check for Cassandra version older than 2.2.x
+                    if (predicateString.equals("Not Supported")) {
+                        break;
+                    }
+                    clusteringColumnSql.add(predicateString);
+                    domainsBuilder.put(columnHandle, domain);
+                    // Check for last clustering column should be only restricted by range condition
+                    if (predicateString.contains(">") || predicateString.contains("<")) {
+                        break;
```

Comment:
Does this actually prevent the case when columns other than the last column have a predicate with `IN`? I think it will just break the loop, and the predicateString would still be constructed. (also, you may want to add a condition for checking whether Cassandra version is < 2.2 when checking for this restriction)